packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -48,18 +49,58 @@ function findAllHotswappableChanges(
   stackChanges.resources.forEachDifference((logicalId: string, change: cfn_diff.ResourceDifference) => {
+    const nonHotswappableResourceFound = isNonHotswappableResourceChange(change);
     const lambdaFunctionShortCircuitChange = isHotswappableLambdaFunctionChange(logicalId, change, assetParamsWithEnv);
-    if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) {
+    const stepFunctionShortCircuitChange = isHotswappableStepFunctionChange(logicalId, change, assetParamsWithEnv);
+
+    const requiresFullDeployment = (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) ||
+                                   (stepFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) ||
+                                   (nonHotswappableResourceFound === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT);
+
+    const irrelevant = (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) &&
+                       (stepFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) &&
+                       (nonHotswappableResourceFound === ChangeHotswapImpact.IRRELEVANT);
+
+    if (requiresFullDeployment) {
       foundNonHotswappableChange = true;
-    } else if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) {
+    } else if (irrelevant) {
       // empty 'if' just for flow-aware typing to kick in...
     } else {
-      hotswappableResources.push(lambdaFunctionShortCircuitChange);
+      if (typeof lambdaFunctionShortCircuitChange !== 'string') {
+        hotswappableResources.push(lambdaFunctionShortCircuitChange);
+      }
+
+      if (typeof stepFunctionShortCircuitChange !== 'string') {
+        hotswappableResources.push(stepFunctionShortCircuitChange);
+      }
     }
   });
   return foundNonHotswappableChange ? undefined : hotswappableResources;
 }
 
+/**
+ * returns `ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT` if a resource was deleted, or a change that we cannot short-circuit occured.
+ * Returns `ChangeHotswapImpact.IRRELEVANT` if a change that does not impact shortcircuiting occured, such as a metadata change.
+ */
+export function isNonHotswappableResourceChange(change: cfn_diff.ResourceDifference): ChangeHotswapImpact {
+  // change.newValue being undefined means the resource was removed; we can't short-circuit that change
+  if (!change.newValue) {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  const newResourceType = change.newValue.Type;
+  // Ignore Metadata changes
+  if (newResourceType === 'AWS::CDK::Metadata') {
+    return ChangeHotswapImpact.IRRELEVANT;
+  }
+  // The only other resource change we should see is a Lambda function or a StepFunctions StateMachine
+  if ((newResourceType !== 'AWS::Lambda::Function') && (newResourceType !== 'AWS::StepFunctions::StateMachine')) {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  return ChangeHotswapImpact.IRRELEVANT;
+}
```

Comment:
I would completely revamp this function. I would make it return a `boolean` instead, and if it returns `false`, I wouldn't even run the other "detector" functions.

It would return `false` if:

* `!change.newValue` (the resource was deleted)
* `!change.oldValue?.Type` (the resource is brand new)
* maybe there are more reasons to return `false` that I'm not seeing right now

In all other cases, we return `true`.

Then, I would add an explicit check for the change being to `AWS::CDK::Metadata` in the caller of this function, and I would again not call the "detector" functions if we find that.

I think that will make the code much more straightforward, and will save us from having to check too many things in the "detector" functions.